src/libs/rust/libc/resources/headers/kdb.h

```diff
@@ -0,0 +1,155 @@
+#include <stdint.h>
+
+enum {
+	ELEKTRA_KEY_KEY_VALUE=1<<1,
+	ELEKTRA_KEY_KEY_FLAGS=3,
+	ELEKTRA_KEY_KEY_BINARY=1<<4,
+	ELEKTRA_KEY_KEY_SIZE=1<<11,
+	ELEKTRA_KEY_KEY_FUNC=1<<12,
+	ELEKTRA_KEY_KEY_META=1<<15,
+	ELEKTRA_KEY_KEY_NULL=1<<16,
+	ELEKTRA_KEY_KEY_LOCK_NAME=1<<17,
+	ELEKTRA_KEY_KEY_LOCK_VALUE=1<<18,
+	ELEKTRA_KEY_KEY_LOCK_META=1<<19,
+};
+
+typedef int elektraLockFlags;
+typedef int elektraKeyFlags;
+
+enum {
+	ELEKTRA_KEY_CP_NAME   = 1<<0,
+	ELEKTRA_KEY_CP_STRING = 1<<1,
+	ELEKTRA_KEY_CP_VALUE  = 1<<2,
+	ELEKTRA_KEY_CP_META   = 1<<3,
+	ELEKTRA_KEY_CP_ALL    = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+};
+typedef unsigned int elektraCopyFlags;
+
+enum {
+	ELEKTRA_NS_NONE=0,
+	ELEKTRA_NS_CASCADING=1,
+	ELEKTRA_NS_META=2,
+	ELEKTRA_NS_SPEC=3,
+	ELEKTRA_NS_PROC=4,
+	ELEKTRA_NS_DIR=5,
+	ELEKTRA_NS_USER=6,
+	ELEKTRA_NS_SYSTEM=7,
+	ELEKTRA_NS_DEFAULT=8,
+};
+typedef int elektraNamespace;
+
+static const elektraNamespace ELEKTRA_NS_FIRST = ELEKTRA_NS_META;
+static const elektraNamespace ELEKTRA_NS_LAST = ELEKTRA_NS_DEFAULT;
+
+typedef ssize_t elektraCursor;
+
+typedef struct _KDB	ElektraKDB;
+typedef struct _Key	ElektraKey;
+typedef struct _KeySet	ElektraKeySet;
+
+/**************************************
+ *
+ * KDB methods
+ *
+ **************************************/
+
+ElektraKDB * elektraKdbOpen (const ElektraKeySet * contract, ElektraKey * parentKey);
+
+int elektraKdbClose (ElektraKDB * handle, ElektraKey * errorKey);
+
+int elektraKdbGet (ElektraKDB * handle, ElektraKeySet * ks, ElektraKey * parentKey);
+int elektraKdbSet (ElektraKDB * handle, ElektraKeySet * ks, ElektraKey * parentKey);
+
+/**************************************
+ *
+ * Key methods
+ *
+ **************************************/
+
+/* Basic Methods */
+ElektraKey * elektraKeyNew (const char * keyname, ...);
+ElektraKey * elektraKeyVNew (const char * keyname, va_list ap);
+
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * source, elektraCopyFlags flags);
+
+int elektraKeyClear (ElektraKey * key);
```

Comment:
@lawli3t what is the latest about the semantics of this function. Probably it should be removed, as we have elektraKeyCopy

Reply:
I wasn't sure, this was included in the header file. I am not aware of any decision we made yet. In my opinion it should be fine to keep actually (even if only as alias) since elektraKeyCopy is not something I would relate to clearing Keys

Reply:
If there is any "slightly different behavior" it is imho problematic and must be documented very well. If you copy a fresh key to an old key, common sense expects the old to be like the fresh key: `elektraKeyClear` shouldn't do anything different.

@lawli3t did you do any investigations into the aliases? We already came to the conclusion that  `elektraKeyClear` is  a good fit to be an alias but afaik we never investigated deeper if aliases are a good idea (if they bring binary+performance benefits worth the hassle to have something which is neither macro nor symbol).

Reply:
It is documented:

https://github.com/ElektraInitiative/libelektra/blob/00138f1fed03ed98c6c2a89b981635f20babaf60/src/libs/elektra/key.c#L283-L286

The part about the reference counter is wrong, since `keyClear` also preserves that. But `keyClear` certainly doesn't respect any of the `KEY_FLAG_RO_*` flags:

https://github.com/ElektraInitiative/libelektra/blob/00138f1fed03ed98c6c2a89b981635f20babaf60/src/libs/elektra/key.c#L541-L567

https://github.com/ElektraInitiative/libelektra/blob/00138f1fed03ed98c6c2a89b981635f20babaf60/src/libs/elektra/key.c#L450-L455

AFAICT `keyClear` is pretty much unused outside of tests, so we could remove it and replace it with

```c
static inline void elektraKeyClear (ElektraKey * key)
{
    elektraKeyCopy (key, NULL, ELEKTRA_KEY_CP_ALL);
}
```

Reply:
> The part about the reference counter is wrong

If you see wrong documentation, please fix it! :rocket: 

> AFAICT keyClear is pretty much unused outside of tests, so we could remove it

I agree.

> and replace it with

Would be interesting to see if the tests continue to work. I don't think elektraKeyClear is an important feature. If someone really needs it, one will figure out you can use elektraKeyCopy. Then you also think about which parts you actually want to have cleared.